initial commit of alliteration filter

ahoimarie opened the pull request:
A simple alliteration filter that checks if a sentence is an alliteration or not.

kaustubhdhole commented on filters/alliteration/README.md:
```diff
@@ -0,0 +1,41 @@
+    * enabled with `strict_polarity = False`
+    * default value for the allowed difference in polarity scores is 0.5; to change it, set the `diff_allowed` argument
+
+Author: Maria Obedkova
```
Please add your email & affiliation.

ahoimarie commented:
Sorry, I did not mean to close the pull request, I wanted to force-reload from the upstream branch to undo my erroneous changes to the wrong folder, which had caused the checks to fail.

james-simon commented:
Simply superb submission! Alliteration's always an amusing attribute to amplify. I just have a few minor comments.

1) In your README, you don't state what your exact criterion for alliterativity is. It's clear from your code, but it's worth mentioning towards the top of the README.

2) That criterion's pretty strong - few people would say that "Peter Piper picked a peck of pickled peppers" **isn't** alliteration! I think it'd make sense to weaken it (e.g. allow one off-word for every $k$ alliterative words) or allow the user to set the filter strength (e.g. take $k$ as an argument).

3) For multi-sentence input, what's your rationale for ignoring all but the first sentence? I'd intuitively expect that the filter would either check whether the entire passage was alliterative on the same letter or whether each sentence is independently alliterative.

Solid work! Let me know if you have any questions.

rteehas commented:
I enjoyed this submission as well. I agree with @james-simon, but on the last point I think that it would make the most sense to treat each sentence independently, since that's probably more likely than an entire passage being alliterative.

mille-s commented:
Thanks for the submission! I think it would be good to add a few words on why sentences with alliterations are a challenge, and for which tasks; right now this is not clear to me (even if I like the idea). As mentioned previously, please also add the method you apply to select the sentences. By the way, this is tagged as "Transformation" but should probably be "Filter".

ahoimarie commented:
Hi @james-simon, 
gratitude and greetings! Alliterations always attract audiences. 

1. I added a few more lines in the README to explain my criterion for alliterativity. Fixed in [9940840](https://github.com/GEM-benchmark/NL-Augmenter/pull/251/commits/994084009326a3287cbfdca22fbb7f07200f7c95)
2. I have added a new variable `min_alliteration_length` to alleviate it. Now, a sentence is deemed an alliteration if it contains at least `min_alliteration_length` words starting with the (first) phoneme found. Would it make sense not to use the first phoneme but the one that occurs most in each sentence?
3. I wanted to avoid checking if a whole paragraph was alliterative because I thought that would practically never be the case. But I agree that ignoring the remaining sentences is not optimal either. I now fixed it to go through all the input sentences and to return true if any of the sentences are alliterative. 

Does this make sense? 
Thanks again!

ahoimarie commented:
Hello @mille-s, 
thanks for your feedback! I have added a few more lines in the README to discuss alliterations and their use. Please see [9940840](https://github.com/GEM-benchmark/NL-Augmenter/pull/251/commits/994084009326a3287cbfdca22fbb7f07200f7c95). Would you like me to expand on it? 
Thanks again!

james-simon commented:
Ahoy! Although arguably ameliorating the aforementioned ails, your changes created certain contrasting concerns:

- if the criterion for alliterativity is for *any* k words to start with the first phoneme, won't a lot of long sentences just satisfy this by accident? For example, if a sentence starts with "The" and includes two other "the"s later, wouldn't that count? This doesn't seem like it actually checks for alliteration; I think the alliterative words do have to be close together. If you agree, you should rethink the rule. As an example, a criterion I might choose would be to take two arguments, `min_alliteration_length` and `allowable_offwords`, and then check if a sentence contains any long run (not necessarily starting with the first word) of `min_alliteration_length` alliterative words, allowing `allowable_offwords` offwords breaking up the run. For example, the sentence `Look, guys: Peter Piper picked a peck of pickled peppers` would satisfy `min_alliteration_length=6` with `allowable_offwords=2` (i.e. "a" and "of"). This criterion seems like it gets much closer to what the typical reader would actually call an alliterative sentence. Does that make sense?
- the header "Why is it a challenge?" doesn't seem right; I'd say something like "Why is this filter important?" instead
- typo in the README: "will they not be removed"

I think checking for a single alliterative sentence makes sense.

mille-s commented:
@ahoimarie thank for the details, I agree it could be used in a generation setting to see if a model manages to create alliterations, although I suspect that the subpopulations with alliterations in the datasets we are using will be extremely small. But with other (more literary) datasets that could be nice.

ahoimarie commented:
Hello! Hopefully having handled your handy help here, I imagine it indeed improved its impact. 

- Instead of checking the first phoneme of the sentence, I now use a rolling window of size `min_alliteration_length+allowable_offwords` to check if it contains one phoneme of at least the frequency `min_alliteration_length` in it. If the sentence is shorter than `min_alliteration_length+allowable_offwords`, the window reduces to `min_alliteration_length`. 
- Agreed, and changed.
- Yes, thanks! 

Thanks! Thorough thoughtful things! 
The thing is, the theme of this thrilling thread thrives without thumbing through a thick thesaurus.

james-simon commented:
Amazing! Apt augmentations, all. Two tiny things:

- I think "If the windowlen is smaller than the length of the data" in your docstring is meaning-reversed
- Is `allowable_offwords` a parameter the user can specify, or is it always 2? It'd be better if it could be chosen

After the aforementioned alterations are achieved, I'll animatedly accept.

ahoimarie commented:
Awesome as always. Appreciated!

- Impressive inspection! I immediately edited it. 
- `allowed_offwords` is indeed a parameter the user can specify. For example `print(Alliteration(stopwords = False, min_alliteration_length=5, allowed_offwords=3).filter("Illuminating illustration of this inane but interesting instruction."))` will return `True` but calling it with `allowed_offwords=2` won't. 

Any additional amendments?

james-simon commented:
Ja, I'm happy! Great filter; every doubt's crushed. Bravo! Accept.

kaustubhdhole commented on Makefile:
```diff
@@ -15,5 +15,5 @@ clean: style
 # Test
 .PHONY: test
-test:
-	pytest -v --cov-config=pyproject.toml
```
This line should not be commented.

ahoimarie replied:
It didn't work on my setup with this config, but changed it back now. Thanks!

kaustubhdhole replied:
Please remove this file from your changes too since backtracking errors would be hard. So, basically this file should not appear in your changes.

ahoimarie replied:
Thanks! I removed the makefile from tracking. Hope this is what you meant?

kaustubhdhole replied:
Hey, @ahoimarie, this file should not be shown as a part of your changes.

kaustubhdhole commented:
All looks good. Please fix the makefile and I am happy to merge :)

kaustubhdhole commented:
The makefile still shows in the tracking as deleted.

ahoimarie commented:
Hi @kaustubhdhole, sorry for the delay, I have been on sick leave because of an injury. 
I've added the Makefile again -- does it work now or shall I create a new PR?

kaustubhdhole commented:
@ahoimarie I missed this message somehow. Please create a separate PR and link this PR in the comment.

kaustubhdhole commented:
Ensure that when you click "Files Changed", the makefile remains untouched.

ahoimarie commented:
@kaustubhdhole I removed the makefile with this little trick mentioned on [stackoverflow](https://stackoverflow.com/questions/39459467/remove-a-modified-file-from-pull-request), as creating a separate PR kept failing. But this should solve it too, shouldn't it?
